rfc/rfc-0000-allow-resources-to-report-status.md

```diff
@@ -0,0 +1,294 @@
+          status: 'Unknown'
+```
+
+```yaml
+---
+apiVersion: carto.run/v1alpha1
+kind: ClusterTemplate
+metadata:
+  name: deploy
+spec:
+  healthyConditionRule:
+    multiMatch:
+      healthy:
+        matchConditions:
+          - type: ReconcileSucceeded
+            status: 'True'
+      unhealthy:
+        matchFields:
+          - key: 'status.conditions[?(@.type=="ReconcileFailed")].status'
+            operator: 'In'
+            values: ['True']
+            messagePath: '.status.usefulErrorMessage'
+
+  template:
+    apiVersion: kappctrl.k14s.io/v1alpha1
+    kind: App
+
+---
+apiVersion: carto.run/v1alpha1
+kind: Workload
+status:
+  resources:
+    - name: deployer
+      conditions:
+        - type: Healthy
+          status: 'False'
+          message: "status.conditions['ReconcileFailed'].status: True: Some kapp useful error message"
+          reason: MatchedField
+          lastTransitionTime: "2022-03-23T13:08:22Z"
+```
+
+```yaml
+---
+apiVersion: carto.run/v1alpha1
+kind: ClusterTemplate
+metadata:
+  name: servicer
+spec:
+  healthyConditionRule:
+    multiMatch:
+      healthy:
+        matchFields:
+          - key: '.status.loadBalancer'
+            operator: 'Exists'
+      unhealthy:
+        matchFields:
+          - key: '.status.loadBalancer'
+            operator: 'DoesNotExist'
+
+  template:
+    apiVersion: v1
+    kind: Service
+
+---
+apiVersion: carto.run/v1alpha1
+kind: Workload
+status:
+  resources:
+    - name: service-creator
+      conditions:
+        - type: Healthy
+          status: 'True'
+          message: "status.loadBalancer: {}"
+          reason: MatchedField
+          lastTransitionTime: "2022-03-23T13:08:22Z"
+```
+
+```yaml
+---
+apiVersion: carto.run/v1alpha1
+kind: ClusterTemplate
+metadata:
+  name: configmap-creator
+spec:
+  healthyConditionRule:
+    alwaysHealthy: {}
+    
+  template:
+    apiVersion: v1
+    kind: ConfigMap
+
+---
+apiVersion: carto.run/v1alpha1
+kind: Workload
+status:
+  resources:
+    - name: configmap-creator
+      conditions:
+        - type: Healthy
+          status: 'True'
+          reason: AlwaysHealthy
+          lastTransitionTime: "2022-03-23T13:08:22Z"
+```
+
+# How it Works
+[how-it-works]: #how-it-works
+
+A stamped resource can be in one of three states:
+1. 'Healthy'
+2. 'Unhealthy'
+3. 'Unknown'
+
+It is up to each template author to determine what "healthy" and "unhealthy" mean for the given resource the template is stamping out. "Unknown" strictly represents that Cartographer has not been able to determine healthy or unhealthy.
+
+Each template will now have a new field in the spec `healthyConditionRule` where authors can specify one of three ways to determine the health of the underlying resource for that template. If no `healthyConditionRule` is defined, Cartographer will default to listing the resource as `Healthy` once it has been successfully applied to the cluster and any relevant outputs have been read off the resource.
```

Comment:
The alternative is to treat templates without a healthyConditionRule as perpetually unknown. While this would technically be more accurate, if we intend to aggregate the resource Healthy conditions into the owner's Ready condition, that could cause a regression.

We could stage the rollout of this behavior, where the initial pass is to treat the default as unknown, and not aggregate it to the owner Ready. In a later release after templates have has a reasonable opportunity to update, then aggregate the value.

Reply:
Interesting idea. This goes to @cirocosta's comment above. In the multi-stage rollout suggestion, are you suggesting we continue to default to unknown in that later release? That would still be a breaking change so just want to make sure to call that out specifically if that's what you're suggesting.